### PR TITLE
Add 16 bit support to kernel operator SOFTMAX

### PR DESCRIPTION
**Pull request by psunn**
This PR is one of steps to extend TFLite to support symmetric 16-bit activations.

In this PR we introduce implementation and tests for SOFTMAX kernel reference function.
The specification of this operator:

SOFTMAX
  Input 0:
    data_type  : int16
    range      : [-32768, 32767]
    granularity: per-tensor, zero_point=0
  Output 0:
    data_type  : int16
    range      : [-32768,32767]
    granularity: per-tensor, zero_point=0

### CODE REVIEW

**Comment by psunn** [comment]
Hi @suharshs @jdduke 
Could you please review this?
As discussed, this one of the INT16 reference kernel reference function.
Thanks!

File: tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h [hunk 1]
```diff
@@ -20,10 +20,21 @@ limitations under the License.
...
 namespace reference_integer_ops {
 
-// Quantized softmax with int8 input and output.
+template<typename T> struct OutputParams {};
+template<> struct OutputParams<int8> {
+  static const int output_num_bits = 8;
+  static const int32 output_min = std::numeric_limits<int8>::min();
+};
+template<> struct OutputParams<int16> {
+  static const int output_num_bits = 16 - 1;
```
Comment (jianlijianli) [inline tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h]: What is the output scale of the int16 softmax?

**Review by jianlijianli — COMMENTED** [review]
Thanks for the PR.

**Comment by gbaned** [comment]
@psunn Can you please address the reviewer comments and resolve conflicts? Thanks!

**Comment by psunn** [comment]
Hi @gbaned , thanks for remind me.
I will resolve it soon.

File: tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h [hunk 2]
```diff
@@ -45,8 +56,11 @@ inline void Softmax(const SoftmaxParams& params,
   const int depth =
       MatchingDim(input_shape, trailing_dim, output_shape, trailing_dim);
 
+  int32 min_num = std::numeric_limits<T>::min();
```
[the remark this answers is not in the document]
Comment (psunn) [reply tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h]: done

File: tensorflow/lite/kernels/activations_test.cc [hunk 3]
```diff
@@ -84,6 +84,9 @@ class BaseActivationsOpModel : public SingleOpModel {
       output_ = AddOutput({input.type, {}, 0, 0, 1. / 256});
     } else if (input.type == TensorType_INT8) {
       output_ = AddOutput({TensorType_INT8, {}, 0, 0, 1. / 256, -128});
+    } else if (input.type == TensorType_INT16) {
+      output_ = AddOutput({TensorType_INT16, {}, 0, 0,
+                           1.0f/(std::numeric_limits<int16_t>::max() + 1), 0});
```
[the remark this answers is not in the document]
Comment (psunn) [reply tensorflow/lite/kernels/activations_test.cc]: done

File: tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h (hunk 1 above)
Comment (psunn) [reply tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h]: 1 / pow(2, 15)

**Comment by psunn** [comment]
Sorry, but please don't merge this change for now, thanks.

File: tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h [hunk 4]
```diff
@@ -57,11 +52,15 @@ inline void Softmax(const SoftmaxParams& params,
   const int depth =
       MatchingDim(input_shape, trailing_dim, output_shape, trailing_dim);
 
-  int32 min_num = std::numeric_limits<T>::min();
-  int32 max_num = std::numeric_limits<T>::max();
-  OutputParams<T> output_params;
+  int32 min_num = std::is_same<InputT, int16_t>::value
```
Comment (sicong-li-arm) [inline tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h]: Minimum value for int16 input can still be negative. I think we should separate input min_num from output min_num.

File: tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h [hunk 5]
```diff
@@ -57,11 +52,15 @@ inline void Softmax(const SoftmaxParams& params,
...
+                      ? 0
+                      : std::numeric_limits<OutputT>::min();
+  int32 max_num = std::numeric_limits<OutputT>::max();
+  int output_num_bits =
+      std::is_same<InputT, int16_t>::value ? 15 : (sizeof(OutputT) * 8);
```
Comment (sicong-li-arm) [inline tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h]:
```suggestion
      std::is_same<OutputT, int16_t>::value ? 15 : (sizeof(OutputT) * 8);
```

File: tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h (hunk 4 above)
Comment (psunn) [reply tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h]: int16 input range [-32767, 32768]

this kernel cover two case:
- softmax with int8 input and int8/int16 output.
- softmax with int16 input and int16 output, zero_point=0.

_std::is_same<InputT, int16_t>::value_ is for distinguish these two case

File: tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h (hunk 5 above)
Comment (psunn) [reply tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h]: same reason as above

File: tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h (hunk 4 above)
Comment (psunn) [reply tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h]: int16 input range [-32767, 32768]

this kernel cover two case:
- softmax with int8 input and int8/int16 output.
- softmax with int16 input and int16 output, zero_point=0.

_std::is_same<InputT, int16_t>::value_ is for distinguish these two case

File: tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h (hunk 5 above)
Comment (psunn) [reply tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h]: thanks for review and discussion.

line 63 should be: InputT max_in_row = std::numeric_limits<InputT>::min();
10:45
Current code is a bit misleading, I'll add some comments.

**Comment by jdduke** [comment]
Just to confirm, this is ready for review?

**Comment by psunn** [comment]
> Just to confirm, this is ready for review?

Hi Jared, no it's still WIP.
alternative implementation is still under discussion internally. Hopefully will update soon.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by psunn** [comment]
@tensorflowbutler
I'm currently working on this PR, should be updated very soon.
Thanks!

**Comment by psunn** [comment]
This PR is ready to review, please have a look, thanks!
@jdduke @suharshs @jianlijianli

**Review by jianlijianli — DISMISSED** [review]
Thanks for the PR. Please take a look at a few minor comments.

File: tensorflow/lite/kernels/internal/softmax_quantized_test.cc [hunk 6]
```diff
@@ -139,14 +198,14 @@ void RunOneSoftmaxTest(const uint8* input_data,
   reference_ops::Softmax(params, shape_common, input_data, shape_common,
                          reference_quant_softmax_output.data());
 
-  CheckOutputData<uint8_t>(optimized_softmax_output.data(),
-                           reference_float_softmax_output.data(), shape_common,
+  CheckOutputData<uint8>(optimized_softmax_output.data(),
```
[the remark this answers is not in the document]
Comment (psunn) [reply tensorflow/lite/kernels/internal/softmax_quantized_test.cc:201]: this function using both uint8_t and uint8, just want to keep them same.

File: tensorflow/lite/kernels/activations.cc [hunk 7]
```diff
@@ -562,6 +568,28 @@ TfLiteStatus SoftmaxPrepare(TfLiteContext* context, TfLiteNode* node) {
...
+    TF_LITE_ENSURE_EQ(context, output->params.zero_point, 0);
+
+    data->params.exp_lut = data->exp_lut;
+    // exp LUT only used on nagative values
+    // we consider exp(-10.0) is insignificant to accumulation
+    gen_lut([](double value) { return std::exp(value); }, -10.0, 0.0,
+            data->params.exp_lut, 513);
+    data->params.one_over_one_plus_x_lut = data->one_over_one_plus_x_lut;
+    gen_lut([](double value) { return 1.0 / (1.0 + value); }, 0.0, 1.0,
+            data->params.one_over_one_plus_x_lut, 513);
```
[the remark this answers is not in the document]
Comment (psunn) [reply tensorflow/lite/kernels/activations.cc:581]: done

File: tensorflow/lite/kernels/internal/softmax_quantized_test.cc [hunk 8]
```diff
@@ -101,18 +120,58 @@ void CheckOutputData(const T* test_output, const T* reference_output,
   }
 }
 
+const float kQuantizedToleranceInt16 = 2 * (1. / 4096);
```
[the remark this answers is not in the document]
Comment (psunn) [reply tensorflow/lite/kernels/internal/softmax_quantized_test.cc]: It was align with [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/activations_test.cc#L163).

I noticed int8 tests has been removed from this file, updated the same for int16.

File: tensorflow/lite/kernels/activations.cc [hunk 9]
```diff
@@ -945,6 +973,24 @@ TfLiteStatus SoftmaxQuantized(TfLiteContext* context, const TfLiteTensor* input,
...
+                                     TfLiteTensor* output,
+                                     SoftmaxOpData* data) {
+  if (NumDimensions(input) >= 1 && NumDimensions(input) <= 4) {
+    reference_ops::SoftmaxInt16(
+        data->params, GetTensorShape(input), GetTensorData<int16_t>(input),
+        GetTensorShape(output), GetTensorData<int16_t>(output));
+    return kTfLiteOk;
+  } else {
+    context->ReportError(
+        context, "Only 1D, 2D, 3D and 4D tensors supported currently, got %dD.",
```
[the remark this answers is not in the document]
Comment (psunn) [reply tensorflow/lite/kernels/activations.cc]: done

**Comment by psunn** [comment]
I update with minor changes to addressing the comments.
Please have a review, and hopefully re-approve if no issues, thanks.
@jianlijianli

**Review by jianlijianli — DISMISSED** [review]
Thanks for making the changes.

File: tensorflow/lite/kernels/activations.cc [hunk 10]
```diff
@@ -945,6 +974,25 @@ TfLiteStatus SoftmaxQuantized(TfLiteContext* context, const TfLiteTensor* input,
...
+                                            const TfLiteTensor* input,
+                                            TfLiteTensor* output,
+                                            SoftmaxOpData* data) {
+  if (NumDimensions(input) >= 1 && NumDimensions(input) <= 4) {
+    reference_ops::SoftmaxInt16(
+        data->params, GetTensorShape(input), GetTensorData<int16_t>(input),
+        GetTensorShape(output), GetTensorData<int16_t>(output));
+    return kTfLiteOk;
+  } else {
+    context->ReportError(context,
```
Comment (rthadur) [inline tensorflow/lite/kernels/activations.cc:988]: `TF_LITE_KERNEL_LOG should be used instead of context->ReportError, so log strings can be removed to save space if needed.`

**Review by rthadur — CHANGES_REQUESTED** [review]
@psunn can you please fix below error ?

File: tensorflow/lite/kernels/activations.cc (hunk 10 above)
Comment (psunn) [reply tensorflow/lite/kernels/activations.cc]: Done. Please have a look.

**Comment by psunn** [comment]
Updated to addressing comments. Please re-approve if no issues.
@rthadur @jianlijianli , thanks.

**Review by jdduke — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

**Review by jianlijianli — APPROVED** [review]

File: tensorflow/lite/kernels/internal/reference/softmax.h [hunk 11]
```diff
@@ -142,6 +142,79 @@ inline void Softmax(const SoftmaxParams& params,
...
+    int32_t shifted_sum =
+        ((static_cast<int64_t>(sum_of_exps) << (headroom_plus_one - 1)) +
+         (1 << 13)) >>
+        14;
+    // since the LUT computes 1/(1 + x) we need to first compute x = (sum - 1).
+    // also, the LUT expects a symmetrical input, so we must also recenter x
+    // from [0, 65535] to [-32768, 32767].
+    int32_t sym_shifted_sum = shifted_sum + (-((1 << 15) + (1 << 16)));
+    int16_t sat_sym_shifted_sum =
+        std::min(std::max(sym_shifted_sum, -32768), 32767);
```
Comment (jdduke) [inline tensorflow/lite/kernels/internal/reference/softmax.h:199]: The internal compiler is complaining with this line:

```
error: no matching function for call to 'max(int32_t&, int)'
           std::min(std::max(sym_scaled_diff, -32768), 32767);
deduced conflicting types for parameter 'const _Tp' ('long int' and 'int')
           std::min(std::max(sym_scaled_diff, -32768), 32767);
```

I think you might need to be explicit in the provided constant types.

**Comment by jdduke** [comment]
Actually, there are a number of internal failures, I'll try to fix them manually.

**Comment by Abhipray** [comment]
@psunn Looks like this PR breaks my application because std::round is not available on my system. Can we use TfLiteRound as defined in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/cppmath.h ?

**Comment by psunn** [comment]
> Actually, there are a number of internal failures, I'll try to fix them manually.

Thanks for the fix them!

**Comment by psunn** [comment]
> @psunn Looks like this PR breaks my application because std::round is not available on my system. Can we use TfLiteRound as defined in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/cppmath.h ?

I raised a [PR](https://github.com/tensorflow/tensorflow/pull/38604), does it solve the problem?@Abhipray

**Comment by Abhipray** [comment]
> > @psunn Looks like this PR breaks my application because std::round is not available on my system. Can we use TfLiteRound as defined in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/cppmath.h ?
> 
> I raised a [PR](https://github.com/tensorflow/tensorflow/pull/38604), does it solve the problem?@Abhipray

Yes, thank you!
